qpmodel/PhysicalNode.cs

```diff
@@ -1904,8 +1907,47 @@ public PhysicNode EmulateSerialization(PhysicNode node)
             var newnode = node.Clone();
             return newnode;
         }
-        public virtual string OpenConsumer(ExecContext context) => null;
-        public virtual string OpenProducer(ExecContext context) => null;
+        public virtual string OpenConsumer(ExecContext econtext)
```

Comment:
I would suggest you keep OpenConsumer() as it is but provides an helper function say createProducersAndChannels() for Redistribute/Broadcast to reuse.

Same as other virtual functions.

Reply:
got it